BigDecimal field values must be rounded according to field format #968

Gavrilov-Ivan opened the pull request:
https://github.com/jmix-framework/jmix/issues/968
Pay close attention to property - is it ok to put it in 'jmix.core'?

gorbunkov commented:
@Gavrilov-Ivan Did you check whether this problem is also relevant for Double or Float datatypes?

Gavrilov-Ivan commented:
Yes, it's relevant for Double and Float datatypes also.
Modified parse process of them the similar way:

1. Parse raw string value as BigDecimal
2. Perform rounding if necessary
3. Check if parsed value fits value-range of target datatype.
4. Get result value as a target datatype

Added tests.

gorbunkov commented on jmix-core/core/src/test/groovy/properties/TestCoreProperties.java:
```diff
@@ -0,0 +1,204 @@
+import java.time.Duration;
+import java.util.List;
+
+public class TestCoreProperties extends CoreProperties {
```
It seems that "groovy" directory is a wrong place for java class. `java/test_support/datatypes` seems to be a better choice.
Also, what is the purpose of the class? Is it for setting the `decimalValueRoundByFormat` property only? If so, then maybe we could add the `@TestPropertySource` annotations to test classes with explicit property value instead?

Gavrilov-Ivan replied:
Moved class to `java/test_support`, without `datatypes` sub-package - this class can be used not only for `datatype` scope of tests.
This class helps to change property values in different tests within single test class. Adding a 'static' property source for entire class looks less suitable for this case. Don't want to split it into multiple test classes based on possible values of involved properties.

gorbunkov commented on jmix-core/core/src/main/java/io/jmix/core/metamodel/datatype/impl/DoubleDatatype.java:
```diff
@@ -82,10 +90,39 @@ public Double parse(String value, Locale locale) throws ParseException {
+        }
+
+        if (!isInDoubleRange(result)) {
+            throw new ParseException(String.format("Double range exceeded: \"%s\"", value), 0);
```
I'm not sure that "exceeded" is a proper word if we speak about negative values. Maybe we can say something like "The value is out of Double datatype range"? The same for Float.

Gavrilov-Ivan replied:
Done

gorbunkov commented on jmix-core/core/src/main/java/io/jmix/core/CoreProperties.java:
```diff
@@ -100,6 +100,12 @@ public class CoreProperties {
+     * Whether BigDecimalDatatype, DoubleDatatype and FloatDatatype should round the actual parsed value
+     * according to format settings.
+     */
+    boolean decimalValueRoundByFormat;
```
Maybe rename the property to `roundDecimalValueByFormat`? WDYT?

Gavrilov-Ivan replied:
Looks a little better - renamed

Gavrilov-Ivan commented on jmix-core/core/src/main/java/io/jmix/core/metamodel/datatype/impl/BigDecimalDatatype.java:
```diff
@@ -94,7 +98,13 @@ public BigDecimal parse(String value, Locale locale) throws ParseException {
         format.setParseBigDecimal(true);
-        return (BigDecimal) parse(value, format);
+        BigDecimal result = (BigDecimal) parse(value, format);
+        if(coreProperties.isBigDecimalValueRoundByFormat()) {
```
Had been done earlier
